### ISSUE TITLE
Make sometimes fails with GMP linking

### ISSUE THREAD

**Issue by certik** [symengine/symengine#367]
As reported by @asmeurer:
```
[ 62%] Building CXX object src/tests/basic/CMakeFiles/test_arit.dir/test_arit.cpp.o
Linking CXX executable test_arit
Undefined symbols for architecture x86_64:
  "operator<<(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, __mpq_struct const*)", referenced from:
      CSymPy::Complex::__str__() const in libcsympy.a(complex.cpp.o)
      CSymPy::Rational::__str__() const in libcsympy.a(rational.cpp.o)
  "operator<<(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, __mpz_struct const*)", referenced from:
      CSymPy::Add::__str__() const in libcsympy.a(add.cpp.o)
      CSymPy::Integer::__str__() const in libcsympy.a(integer.cpp.o)
      operator<<(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::map<std::__1::vector<int, std::__1::allocator<int> >, __gmp_expr<__mpz_struct [1], __mpz_struct [1]>, std::__1::less<std::__1::vector<int, std::__1::allocator<int> > >, std::__1::allocator<std::__1::pair<std::__1::vector<int, std::__1::allocator<int> > const, __gmp_expr<__mpz_struct [1], __mpz_struct [1]> > > > const&) in libcsympy.a(dict.cpp.o)
      operator<<(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::unordered_map<std::__1::vector<int, std::__1::allocator<int> >, __gmp_expr<__mpz_struct [1], __mpz_struct [1]>, CSymPy::vec_int_hash, CSymPy::vec_int_eq, std::__1::allocator<std::__1::pair<std::__1::vector<int, std::__1::allocator<int> > const, __gmp_expr<__mpz_struct [1], __mpz_struct [1]> > > > const&) in libcsympy.a(dict.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/tests/basic/test_arit] Error 1
make[1]: *** [src/tests/basic/CMakeFiles/test_arit.dir/all] Error 2
make: *** [all] Error 2
```
This is a bug in csympy's cmake files, because if there are problems with the GMP library, it should be caught by cmake already, which should report a nice message to the user what the problem is. If cmake succeeds, then make should always work.

We need to figure out what exactly the problem here is, and then add a proper check to cmake for this.

**Comment by certik** [comment]
I found this: http://stackoverflow.com/questions/24641199/c-gmp-and-homebrew, which seems similar. It looks like GMP might have been compiled with a different compiler or a different standard lib. Hm, what a pain.

**Comment by certik** [comment]
Maybe the issue #344 is relevant. 

We really need to get Travis testing on OS X. Then we can use Conda to install GMP and things would be tested that they work.

**Comment by asmeurer** [comment]
My gmp package was probably compiled on OS X 10.7, whereas I am trying to compile csympy on OS X 10.10.

**Comment by certik** [comment]
That could do it. Try to recompile gmp with the same compiler.

**Comment by isuruf** [comment]
@asmeurer, I get the same error when compiling with `gcc` after linking with a `gmp` library compiled with `AppleClang`

A fix for this issue would be avoid using `ostringstream << mpq_class` and `ostringstream << mpz_class`. For eg:
```
mpz_class a = 4;
std::cout << a;
```
will give this problem, but following won't;
```
mpz_class a = 4;
std::cout << a.get_str();
```

Is this fix okay? or should we disable compiling in a case like this?

**Comment by certik** [comment]
No, this is not the right fix. The problem is that you cannot mix C++
libraries compiled with different compilers and/or different standard
libraries. So even if your fix temporarily works, there is no guarantee it
will always work, nor is the behavior well defined. The fix is to compile
gmp and symengine with the same compiler.

Alternatively we can implement our own wrapper for mpz_t, similarly to what
we do with mpc and mpfr. Then we only need to link with the C gmp
interface, which is ABI compatible across compilers.

But the same problem is there for any other C++ dependency, so ultimately
one has to use the same C++ compiler for everything.

**Comment by isuruf** [comment]
Makes sense. Shall we have a small program that tries to do something like,
```
mpz_class a = 4;
std::cout << a;
```
and give an error message saying gmp library linked is not compatible with the compiler used?

**Comment by certik** [comment]
Yes! That would indeed be awesome.

**Comment by isuruf** [comment]
#476 does the check at configuration time.
#473 tests compiling `symengine` using `g++` with a `gmp` compiled using `g++`.

**Comment by bluescarni** [comment]
@certik clang is binary compatible with GCC, and by default it uses GCC's libstdc++ unless told otherwise (at least on Linux). I always compile Piranha with both GCC and clang, on a Linux box on which everything was compiled with GCC (including Boost and GMP), and I never had a problem.

Issues arise if you mix up standard libraries though. Clang can use another standard library, called libc++, if instructed so. I believe on recent OSX versions clang uses libc++ by default. If you link together libraries compiled with libstdc++ and libc++, you will run into the issues you described above.

To summarize, you have the following cases:

* you compile everything with GCC, no issues here;
* you compile some libraries with GCC, other libraries with clang + libstdc++ (default behaviour in Linux), no issues;
* you compile everything with clang + libc++ (as, I presume, in recent OSX versions), no issue here;
* you mix libraries compiled with GCC and clang + libc++, here you have problems.

**Comment by isuruf** [comment]
Closing this issue as checks are done at cmake time to make sure it works.

Issue was that prior to OS X 10.9, default was libstdc++ and GMP was compiled with it and SymEngine was compiled with libc++ which is the default in OS X 10.9.

**Comment by zhaojunchen** [comment]
Today I have same problem about cin and cout .My pc is macbook air,and IDE  is Clion.
Look the official document about c++ Input module
```
10.3 C++ Formatted Output

The following functions are provided in libgmpxx (see Section 3.1 [Headers and Libraries], page 17), which is built if
 C++ support is enabled (see Section 2.1 [Build Options], page 3). Prototypes are available from <gmp.h>.
```
I don't add lingmpxx moudle at the my CmakeLists.txt  ,so I add it into my project cmakelists.txt as follows:
```
cmake_minimum_required(VERSION 3.13)
project(MyGMP)
set(CMAKE_CXX_STANDARD 11)
add_executable(MyGMP main.cpp)

# To use GMP, you need add four lines into you CMakeLists.txt
# You need find your GMP installed path firstly, and Then add them.

include_directories("/usr/local/include")
include_directories("/usr/local/lib")

target_link_libraries(MyGMP "/usr/local/lib/libgmp.a")
target_link_libraries(MyGMP "/usr/local/lib/libgmpxx.a")
```
The it is ok!
